server/aesdsocket.c

```diff
@@ -147,20 +186,59 @@ int main(int argc, char **argv) {
               addrString, sizeof(connectedAddr));
     syslog(LOG_INFO, "Accepted connection from %s", addrString);
 
+    // DON'T clean up the memory here, since we would destroy it too early
+    struct thread_entry_t *threadTrackingData =
+        malloc(sizeof(struct thread_entry_t));
+    if (threadTrackingData == NULL) {
+      syslog(LOG_ERR, "Could not allocate thread tracking structure");
+      return EXIT_FAILURE;
+    }
+
     // Run the server behavior (read a line, write the file)
-    if(on_server_connection(connectionSocketFd, tmpfile) != EXIT_SUCCESS){
+    if (on_server_connection(connectionSocketFd, tmpfile, &tmpFileMutex,
+                             addrString, &(threadTrackingData->worker_thread),
+                             &(threadTrackingData->thread_complete)) !=
+        EXIT_SUCCESS) {
       syslog(LOG_ERR, "Server processing failed");
+      free(threadTrackingData);
       return EXIT_FAILURE;
     }
 
-    // Connection closed automatically due to scoped cleanup of connectionSocketFd
-    syslog(LOG_INFO, "Closed connection from %s", addrString);
+    SLIST_INSERT_HEAD(&threadList, threadTrackingData, _entry);
+
+    // Clean up existing threads when we make a new one
+    struct thread_entry_t *threadEntry;
+    struct thread_entry_t *lastEntry = NULL;
+    SLIST_FOREACH(threadEntry, &threadList, _entry){
+      if(!atomic_flag_test_and_set(&(threadEntry->thread_complete))){
```

Comment:
@bennowotny this works but is a little confusing since thread_complete is set to 0 when complete.  Since there shouldn't be a need to modify this value outside the thread I think you can remove the atomic logic.  Could also replace with atomic_load/atomic_bool perhaps if you want to keep the atomic.